Validation/RecoParticleFlow/plugins/PFCaloGPUComparisonTask.cc

```diff
@@ -0,0 +1,170 @@
+  MonitorElement* pfCluster_Eta_GPUvsCPU_;
+  MonitorElement* pfCluster_Phi_GPUvsCPU_;
+
+  std::string pfCaloGPUCompDir;
+};
+
+PFCaloGPUComparisonTask::PFCaloGPUComparisonTask(const edm::ParameterSet& conf)
+    : pfClusterTok_ref_{consumes<reco::PFClusterCollection>(
+          conf.getUntrackedParameter<edm::InputTag>("pfClusterToken_ref"))},
+      pfClusterTok_target_{
+          consumes<reco::PFClusterCollection>(conf.getUntrackedParameter<edm::InputTag>("pfClusterToken_target"))},
+      pfCaloGPUCompDir{conf.getUntrackedParameter<std::string>("pfCaloGPUCompDir")} {}
+
+void PFCaloGPUComparisonTask::bookHistograms(DQMStore::IBooker& ibooker,
+                                             edm::Run const& irun,
+                                             edm::EventSetup const& isetup) {
+  const char* histo;
+
+  ibooker.setCurrentFolder("ParticleFlow/" + pfCaloGPUCompDir);
+
+  histo = "pfCluster_Multiplicity_GPUvsCPU_";
+  pfCluster_Multiplicity_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 2000, 100, 0, 2000);
+
+  histo = "pfCluster_Energy_GPUvsCPU_";
+  pfCluster_Energy_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 500, 100, 0, 500);
+
+  histo = "pfCluster_RecHitMultiplicity_GPUvsCPU_";
+  pfCluster_RecHitMultiplicity_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Layer_GPUvsCPU_";
+  pfCluster_Layer_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Depth_GPUvsCPU_";
+  pfCluster_Depth_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Eta_GPUvsCPU_";
+  pfCluster_Eta_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+
+  histo = "pfCluster_Phi_GPUvsCPU_";
+  pfCluster_Phi_GPUvsCPU_ = ibooker.book2D(histo, histo, 100, 0, 100, 100, 0, 100);
+}
+void PFCaloGPUComparisonTask::analyze(edm::Event const& event, edm::EventSetup const& c) {
+  edm::Handle<reco::PFClusterCollection> pfClusters_ref;
+  event.getByToken(pfClusterTok_ref_, pfClusters_ref);
+
+  edm::Handle<reco::PFClusterCollection> pfClusters_target;
+  event.getByToken(pfClusterTok_target_, pfClusters_target);
+
+  //
+  // Compare per-event PF cluster multiplicity
+
+  if (pfClusters_ref->size() != pfClusters_target->size())
+    LOGVERB("PFCaloGPUComparisonTask") << " PFCluster multiplicity " << pfClusters_ref->size() << " "
+                                       << pfClusters_target->size();
+  pfCluster_Multiplicity_GPUvsCPU_->Fill((float)pfClusters_ref->size(), (float)pfClusters_target->size());
+
+  //
+  // Find matching PF cluster pairs
+  std::vector<int> matched_idx;
+  for (unsigned i = 0; i < pfClusters_ref->size(); ++i) {
+    bool matched = false;
+    for (unsigned j = 0; j < pfClusters_target->size(); ++j) {
+      if (pfClusters_ref->at(i).seed() == pfClusters_target->at(j).seed()) {
+        if (!matched) {
+          matched = true;
+          matched_idx.push_back((int)j);
+        } else {
```

Comment:
If you don't expect multiple matches, consider just breaking after you find a match to save some computation time.  If you consider this LogWarning is important for catching unexpected behavior, disregard.